Update CmdStan read and logic

ahartikainen opened the pull request:
## Description
Read CmdStan csv files manually. This enables us to parse large models (100k parameters) much faster than pandas.

This PR also adds `dtypes` argument, which user can use to transform dtypes for specific parameters

    dtypes = {"theta": int}

## Checklist
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md#pull-request-checklist) PR format
- [ ] Includes a sample plot to visually illustrate the changes (only for plot-related functions)
- [ ] New features are properly documented (with an example if appropriate)?
- [ ] Includes new or updated tests to cover the new feature
- [x] Code style  correct (follows pylint and black guidelines)
- [ ] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/main/CHANGELOG.md#v0xx-unreleased)

<!--
Also, please consider reading the contributing guidelines and code of conduct carefully before submitting the PR. They are available at
- https://github.com/arviz-devs/arviz/blob/main/CONTRIBUTING.md
- https://github.com/arviz-devs/arviz/blob/main/CODE_OF_CONDUCT.md

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in. Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

OriolAbril commented:
I thought that pandas reader was written in C to be fast, do you have some advise on which cases make having a custom worth it?

ahartikainen commented:
I will give you example later today.

ahartikainen commented:
```
import tempfile
import numpy as np
import pandas as pd
from pathlib import Path
from uuid import uuid4
import time
import pandas as pd
import numpy as np
import shutil
import matplotlib.pyplot as plt

def read_output_file_manual(path):
    comments = []
    data = []
    columns = None
    with open(path, "rb") as f_obj:
        # read header
        for line in f_obj:
            if line.startswith(b"#"):
                comments.append(line.decode("utf-8").strip())
                continue
            columns = {key: idx for idx, key in enumerate(line.strip().decode("utf-8").split(","))}
            break

        for line in f_obj:
            line = line.strip()
            if line.startswith(b"#"):
                comments.append(line.decode("utf-8"))
                continue
            if line:
                data.append(line.split(b","))

        data = np.array(data, dtype=np.float64)

    return columns, data, comments

def read_output_file_pandas(path):
    comments = []
    data = []
    columns = None
    with open(path, "rb") as f_obj:
        # read header
        for line in f_obj:
            if line.startswith(b"#"):
                comments.append(line.decode("utf-8").strip())
                continue
            columns = {key: idx for idx, key in enumerate(line.strip().decode("utf-8").split(","))}
            break

        f_obj_loc = f_obj.tell()
        for line in f_obj:
            if line.startswith(b"#"):
                comments.append(line.strip().decode("utf-8"))
                continue
        f_obj.seek(f_obj_loc)
        data = pd.read_csv(f_obj, header=None, comment="#", float_precision="high", dtype=np.float64).values

    return columns, data, comments


%%time
np.random.seed(10)
reference_files = {}
with tempfile.TemporaryDirectory() as tmpdir:
    path = Path(tmpdir)
    for parameter_size in [10, 100, 1000, 10_000, 20_000, 50_000]:
        for draw_size in [100, 500, 1000, 10_000]:
            data = np.random.randn(draw_size, parameter_size)
            columns = [str(uuid4()) for _ in range(parameter_size)]
            df = pd.DataFrame(data=data, columns=columns)
            output_path = Path(tmpdir) / f"parameters_{parameter_size}_draws_{draw_size}.csv"
            output_path2 = Path(tmpdir) / f"parameters_{parameter_size}_draws_{draw_size}_2.csv"
            df.to_csv(str(output_path))
            shutil.copy(output_path, output_path2)
            reference_files[(parameter_size, draw_size)] = (output_path, output_path2)
    
    results = {"pandas": {}, "manual": {}}
    for key, (path, path2) in reference_files.items():
        res_manual = []
        res_pandas = []
        for n in range(1):
            st = time.time()
            val = read_output_file_manual(path)
            et = time.time()
            res_manual.append(et - st)
            
            st = time.time()
            val = read_output_file_pandas(path2)
            et = time.time()
            res_pandas.append(et - st)
            
        results["manual"][key] = res_manual
        results["pandas"][key] = res_pandas

res_df = pd.DataFrame(results).applymap(lambda x: np.mean(x)).reset_index().rename(columns={"level_0": "parameters", "level_1": "draws"})

fig, ax = plt.subplots(1, dpi=100, figsize=(7,7))
for i, (group, gdf) in enumerate(res_df.groupby(by="parameters")):
    plt.plot(gdf["draws"], gdf["pandas"], marker='.', label=f"pandas par:{group}", c=f"C{i%8}")
    plt.plot(gdf["draws"], gdf["manual"], marker='.', label=f"manual par:{group}", c=f"C{i%8}", ls="--")
    
plt.xlabel("Parameters")
plt.ylabel("Timing (s)")
plt.legend()
plt.savefig("./csv_read_comparison.png", dpi=200, bbox_inches="tight")
```

I will upload comparison image after I have run the script

ahartikainen commented:
# Results

<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>parameters</th>\n      <th>draws</th>\n      <th>pandas</th>\n      <th>manual</th>\n      <th>numpy</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>2</th>\n      <td>1000</td>\n      <td>10000</td>\n      <td>2.103543</td>\n      <td>4.698918</td>\n      <td>11.429663</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>50000</td>\n      <td>10000</td>\n      <td>420.375299</td>\n      <td>227.831740</td>\n      <td>1131.707759</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>100000</td>\n      <td>10000</td>\n      <td>1750.873820</td>\n      <td>1399.882980</td>\n      <td>5438.172931</td>\n    </tr>\n  </tbody>\n</table>

# Difference against pandas (x - pandas)

<table border="1" class="dataframe">\n  <thead>\n    <tr style="text-align: right;">\n      <th></th>\n      <th>parameters</th>\n      <th>draws</th>\n      <th>manual</th>\n      <th>numpy</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>2</th>\n      <td>1000</td>\n      <td>10000</td>\n      <td>2.595374</td>\n      <td>9.326120</td>\n    </tr>\n    <tr>\n      <th>5</th>\n      <td>50000</td>\n      <td>10000</td>\n      <td>-192.543559</td>\n      <td>711.332459</td>\n    </tr>\n    <tr>\n      <th>8</th>\n      <td>100000</td>\n      <td>10000</td>\n      <td>-350.990840</td>\n      <td>3687.299111</td>\n    </tr>\n  </tbody>\n</table>


![csv_read_comparison](https://user-images.githubusercontent.com/13161958/108226363-8439e180-7145-11eb-84a0-c126bd2b46ad.png)


I don't know, maybe go with the manual handling? All csv files should be "good" (we don't need to consider ill-formed ones)

ahartikainen commented:
@OriolAbril are you happy with the changes?

Code should now a bit more clear than previously.

OriolAbril commented:
Looks good, thanks!
